syntax = "proto2";

package eb;

message Size2i {
	required uint32 width = 1;
	required uint32 height = 2;
}

message Point2i {
	required int32 x = 1;
	required int32 y = 2;
}

message Size2f {
	required float width = 1;
	required float height = 2;
}

message Rect2i {
	required int32 left = 1;
	required int32 top = 2;
	required int32 right = 3;
	required int32 bottom = 4;
}

message Colour {
	required float r = 1;
	required float g = 2;
	required float b = 3;
}

enum EditMode {
	EM_FREE = 0;
	EM_CORNER_AVERAGE = 1;
	EM_ROW_AVERAGE = 2;
	EM_COLUMN_AVERAGE = 3;
	EM_EDGE_AVERAGE = 4;
	//EM_WATERSHED_AVERAGE = 5;
	//EM_PERSPECTIVE = 6;
}

enum DisplayMode {
	DM_MESH_GRID = 0;
	DM_PURE_COLOUR = 1;
}

enum SelectMode {
	SM_SINGLE = 0x00;
	SM_ROW = 0x01;
	SM_COLUMN = 0x02;
	SM_CROSS = 0x03;
}

message EditorSettings {
	required DisplayMode display_mode = 12;
	required EditMode edit_mode = 1;
	required Colour line_colour = 4;
	required Colour background_colour = 5;
	required float adjust_speed = 7;
	required bool show_control_point = 8;
	required Size2i grid_size = 9;
	required SelectMode select_mode = 10;
	required Point2i selected = 11;
}

enum ControlPointType {
	CP_FREE = 0x00;
	CP_CEASE = 0x01;
}

message ControlPoint {
	required ControlPointType type = 1;
	required float x = 2;
	required float y = 3;
}

message LatticeSettings {
	required Size2i size = 1;
	required uint32 subdiv_level = 2;
	repeated ControlPoint control_points = 3;
}

message EdgeBlendSettings {
	required bool enable = 1;
	required uint32 width = 2;
	required float blending = 3;
	required float gamma = 4;
	required float center = 5;
}

message Workspace {
	required EditorSettings editor_settings = 1;
	required LatticeSettings lattice_settings = 2;
	required EdgeBlendSettings left_edge_settings = 3;
	required EdgeBlendSettings top_edge_settings = 4;
	required EdgeBlendSettings right_edge_settings = 5;
	required EdgeBlendSettings bottom_edge_settings = 6;
}